enhancements/machine-config/admin-defined-node-disruption-policy.md

```diff
@@ -0,0 +1,311 @@
+ - Draining a node
+ - Reloading a service
+ - Daemon-reload and restarting of a service
+ - Reboot (default)
+ - Special (see workflow description)
+
+In the future, we can extend this in two ways:
+Supporting more changes/actions (e.g. user defined script to run after a certain change)
+Supporting diffing between ostree images (for image based updates) and bootc managed systems (for bifrost)
+
+
+### Workflow Description
+
+An admin of an OCP cluster can at any time create/update/delete nodeDisruptionPolicySpec fields in the MachineConfiguration object to reconfigure how they wish updates to be applied. There isn’t any immediate effect so long as the CR itself is valid.
+
+The MachineConfigController’s drain controller will process any update to the object and validate the NodeDisruptionPolicy for duplication, action validity (but not correctness), etc. If the object is valid, the drain controller will update the object’s status to indicate success, and if not, update those and degrade the controller, example:
+
+```
+status:
+  validation: Failed
+  Reason: Invalid NodeDisruptionPolicy spec: xxx
+```
+
+The MachineConfigDaemon will process and apply the current status.nodeDisruptionPolicyStatus.clusterPolicies, which holds the previous validated object. The spec changes will be ignored until validated and populated by the controller.
+
+When the user first enables the new nodeDisruptionPolicySpec via featuregate (or at some version, by default), the "default" MachineConfiguration object will be generated with the existing cluster defaults by the MachineConfigController, so the user can also view the default policies that came with the cluster. The user can update these fields if they wish to use a different workflow via userPolicies, but are not allowed to modify clusterDefaultPolicies, which is reserved for MCO to update these as we wish in the future. The controller will enforce this when generating the status. The fields that exist today are as highlighted above.
+
+Note that /etc/containers/registries.conf has a “Special” keyword. This is due to the extra processing the MCO does today. No user defined policy may use the “Special” keyword.
+
+User-provided changes will overwrite the cluster defaults, reflected in the status.
+
+The special MCO cert paths ("/etc/kubernetes/kubelet-ca.crt", "/etc/kubernetes/static-pod-resources/configmaps/cloud-config/ca-bundle.pem") are not listed in the clusterDefaultPolicies as they are no longer MachineConfig changes.
+
+In terms of the policy itself, in case a change has multiple matching policies (e.g. two files changed, both with a different policy), all policies will be applied in order (e.g. reload multiple services). If any change in the list does not match any defined policy, the MCO will default to rebooting and skip over all the policies.
+
+The MachineConfigNodes CR will track the status of the upgrade, and will also be updated to indicate what the NodeDisruptionPolicy used for the update was.
```

Comment:
Do we have an API extension for this?

Reply:
Not today, since MCN is also tech preview.

Will make it a goal for GA